src/unixfs-data-format.md

```diff
@@ -0,0 +1,652 @@
+[`raw` block](#raw-blocks).
+
+For example, consider this pseudo-json block:
+
+```json
+{
+  "Links": [{"Hash":"Qmfoo"}, {"Hash":"Qmbar"}],
+  "Data": {
+    "Type": "File",
+    "blocksizes": [20, 30]
+  }
+}
+```
+
+This indicates that this file is the concatenation of the `Qmfoo` and `Qmbar` files.
+
+When reading a file represented with `dag-pb`, the `blocksizes` array gives us the
+size in bytes of the partial file content present in children DAGs. Each index in
+`PBNode.Links` MUST have a corresponding chunk size stored at the same index
+in `decode(PBNode.Data).blocksizes`.
+
+:::warning
+Implementers need to be extra careful to ensure the values in `Data.blocksizes`
+are calculated by following the definition from [`Blocksize`](#decodepbnodedatablocksize).
+:::
+
+This allows for fast indexing into the file. For example, if someone is trying
+to read bytes 25 to 35, we can compute an offset list by summing all previous
+indexes in `blocksizes`, then do a search to find which indexes contain the
+range we are interested in.
+
+In the example above, the offset list would be `[0, 20]`. Thus, we know we only need to download `Qmbar` to get the range we are interested in.
+
+UnixFS parser MUST error if `blocksizes` or `Links` are not of the same length.
+
+#### `decode(PBNode.Data).Data`
+
+An array of bytes that is the file content and is appended before
+the links. This must be taken into account when doing offset calculations; that is,
+the length of `decode(PBNode.Data).Data` defines the value of the zeroth element
+of the offset list when computing offsets.
+
+#### `PBNode.Links[].Name`
+
+This field makes sense only in :ref[Directories] contexts and MUST be absent
+when creating a new file. For historical reasons, implementations parsing
+third-party data SHOULD accept empty values here.
+
+If this field is present and non-empty, the file is invalid and the parser MUST
+error.
+
+#### `decode(PBNode.Data).Blocksize`
+
+This field is not directly present in the block, but rather a computable property
+of a `dag-pb`, which would be used in the parent node in `decode(PBNode.Data).blocksizes`.
+It is the sum of the length of `decode(PBNode.Data).Data` field plus the sum
+of all link's `blocksizes`.
+
+#### `decode(PBNode.Data).filesize`
+
+If present, this field MUST be equal to the `Blocksize` computation above.
+Otherwise, this file is invalid.
+
+#### `dag-pb` `File` Path Resolution
+
+A file terminates a UnixFS content path. Any attempt to resolve a path past a
+file MUST error.
+
+### `dag-pb` `Directory`
+
+A :dfn[Directory], also known as folder, is a named collection of child :ref[Nodes]:
+
+- Every link in `PBNode.Links` is an entry (child) of the directory, and
+  `PBNode.Links[].Name` gives you the name of that child.
+- Duplicate names are not allowed. Therefore, two elements of `PBNode.Link` CANNOT
+  have the same `Name`. If two identical names are present in a directory, the
+  decoder MUST fail.
```

Comment:
```suggestion
- Duplicate names within a given directory are not allowed. Therefore, two elements
  of `PBNode.Link` CANNOT have the same `Name`. If two identical names are present in
  the same a directory, the directory must be consider malformed and the decoder MUST
  fail.
```
"identical" means stringwise/semantically but not bytewise, right? Practically it is probably a moot distinction but just wondering because the type of people who make their own parsers at this low a level might care for pedantic, if not practical, reasons.